default/AI/ResourcesAI.py

```diff
@@ -211,30 +376,225 @@ def assess_protection_focus(pid):
+            if focus_manager.bake_future_focus(pid, IFocus):
+                print "%s focus of planet %s (%d) (with Honeycomb) at Industry Focus" % (["set", "left"][pinfo.current_focus == IFocus], planet.name, pid)
+                continue
+        if ((([bld.buildingTypeName for bld in map(universe.getObject, planet.buildingIDs) if bld.buildingTypeName in
+               ["BLD_CONC_CAMP", "BLD_CONC_CAMP_REMNANT"]] != [])
+             or ([ccspec for ccspec in planet.specials if ccspec in
+                  ["CONC_CAMP_MASTER_SPECIAL", "CONC_CAMP_SLAVE_SPECIAL"]] != []))
+                and IFocus in planet.availableFoci):
+            if focus_manager.bake_future_focus(pid, IFocus):
+                if pinfo.current_focus != IFocus:
+                    print ("Tried setting %s for Concentration Camp planet %s (%d) with species %s and current focus %s, got result %d and focus %s" %
+                           (pinfo.future_focus, planet.name, pid, planet.speciesName, pinfo.current_focus, True, planet.focus))
+                print "%s focus of planet %s (%d) (with Concentration Camps/Remnants) at Industry Focus" % (["set", "left"][pinfo.current_focus == IFocus], planet.name, pid)
+                continue
+            else:
+                newplanet = universe.getPlanet(pid)
+                print ("Error: Failed setting %s for Concentration Camp planet %s (%d) with species %s and current focus %s, but new planet copy shows %s" %
+                       (pinfo.future_focus, planet.name, pid, planet.speciesName, planet.focus, newplanet.focus))
+
+
+def set_planet_protection_foci(focus_manager):
+    """Assess and set protection foci"""
+    universe = fo.getUniverse()
+    for pid, pinfo in focus_manager.raw_planet_info.items():
+        planet = pinfo.planet
+        if PFocus in planet.availableFoci and assess_protection_focus(pid, pinfo):
+            curFocus = planet.focus
+            if focus_manager.bake_future_focus(pid, PFocus):
+                if curFocus != PFocus:
+                    print ("Tried setting %s for planet %s (%d) with species %s and current focus %s, got result %d and focus %s" %
+                           (pinfo.future_focus, planet.name, pid, planet.speciesName, curFocus, True, planet.focus))
+                print "%s focus of planet %s (%d) at Protection(Defense) Focus" % (["set", "left"][curFocus == PFocus], planet.name, pid)
+                continue
+            else:
+                newplanet = universe.getPlanet(pid)
+                print ("Error: Failed setting %s for planet %s (%d) with species %s and current focus %s, but new planet copy shows %s" %
+                       (focus_manager.new_foci[pid], planet.name, pid, planet.speciesName, planet.focus, newplanet.focus))
+
+
+def set_planet_happiness_foci(focus_manager):
+    """Assess and set planet focus to preferred focus depending on happiness"""
+    # TODO Assess need to set planet to preferred focus to improve happiness
+
+
```

Comment:
Comments does not count.  Function have docstring.  It is part of function declaration and can be accessed in runtime via `__doc__ `